nav2_costmap_2d/src/footprint_subscriber.cpp

```diff
@@ -23,49 +23,72 @@ namespace nav2_costmap_2d
 
 FootprintSubscriber::FootprintSubscriber(
   nav2_util::LifecycleNode::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  const double & footprint_timeout)
 : FootprintSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_clock_interface(),
+    topic_name, footprint_timeout)
 {}
 
 FootprintSubscriber::FootprintSubscriber(
   rclcpp::Node::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  const double & footprint_timeout)
 : FootprintSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_clock_interface(),
+    topic_name, footprint_timeout)
 {}
 
 FootprintSubscriber::FootprintSubscriber(
   const rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
   const rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
   const rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
-  const std::string & topic_name)
+  const rclcpp::node_interfaces::NodeClockInterface::SharedPtr node_clock,
+  const std::string & topic_name,
+  const double & footprint_timeout)
 : node_base_(node_base),
   node_topics_(node_topics),
   node_logging_(node_logging),
-  topic_name_(topic_name)
+  node_clock_(node_clock),
+  topic_name_(topic_name),
+  footprint_timeout_(rclcpp::Duration(footprint_timeout, 0.))
 {
   footprint_sub_ = rclcpp::create_subscription<geometry_msgs::msg::PolygonStamped>(node_topics_,
       topic_name, rclcpp::SystemDefaultsQoS(),
       std::bind(&FootprintSubscriber::footprint_callback, this, std::placeholders::_1));
 }
 
 bool
-FootprintSubscriber::getFootprint(std::vector<geometry_msgs::msg::Point> & footprint)
+FootprintSubscriber::getFootprint(
+  std::vector<geometry_msgs::msg::Point> & footprint,
+  rclcpp::Time & stamp)
 {
   if (!footprint_received_) {
     return false;
   }
 
   footprint = toPointVector(
     std::make_shared<geometry_msgs::msg::Polygon>(footprint_->polygon));
+  auto & footprint_stamp = footprint_->header.stamp;
+
+  if (stamp - footprint_stamp > footprint_timeout_) {
```

Comment:
This logic makes sense now, but I still am unsure whether we should to let the user have the time stamp of the latest footprint and choose what to do (I believe that's what the previous PR was doing) OR as you've done here have the user choose a timeout upon creation of this subscriber class. What if they had different timeouts for different usages, they'd need to create a new object?

Reply:
I don't think there's a situation when you'd have 1 application that cares about having a 1s vs a 0.5s timeout where you'd need to do that. However if one were to appear, I think it would be totally reasonable to create get/set methods to change the timeout, or include  the timeout on calling. I just didn't want to over complicate it without a compelling reason. 

Even if the method return false that its an old footprint, the user still gets the footprint intentionally if it chooses to still use it. I think it makes sense to encapsulate the timeout in the subscriber utility. I would generally agree if this was the static footprint (ei this was just the shape of the robot in some normalize coodinates) but since its the posed-footprint, its critical you know if that information is stale for collision checking or else you'll be checking for collisions in the pose from `N` seconds ago and not be the wiser. If you call it default you'll get the current footprint at current time, otherwise you can specify another timestamp with the overload to check against. 

Either way, the user will get the last footprint. I'd disagree that it should be on the user of the subscriber to check if its old, I think encapsulating it in the subscriber is a very safe thing to do so a user when creating the object needs to be intentional about safely utilizing the posed footprint info.

Reply:
I just see it as a limiting use case since you can't currently change the timeout, or perhaps there is a need to change it after the object was constructed.

Technically, the footprint is transformed into normalized coordinates in the collision checking by grabbing the current pose and then we transform it given a new pose to check for collisions. I suppose it could be static, but since we don't have direct access to the footprint spec on `Costmap2DROS` and the footprint itself can change I used the subscriber. One consequence I'm now thinking is that the footprint and current pose need to be in sync (i.e. about the same time), so perhaps that needs to be a check in the `collision_checker`. 

https://github.com/ros-planning/navigation2/blob/74dd39fd6932f11abae1a02cbcd68579638b0f47/nav2_costmap_2d/src/collision_checker.cpp#L171

Reply:
So it sounds like you're mostly concerned about not being able to change the timeout on the fly. Would it fix that if I provided a timeout argument to the get footprint method that defaulted to the timeout set on creation? That way it generally works with that timeout unless overloaded?

Reply:
If that’s not really “in the spirit” of what you mean, let me know and lets find another way. I’m just not entirely sure what is the best way to provide what youre asking for cleanly.

Reply:
I think adding the overload would be a good compromise

Reply:
OK done. Unfortunately I couldn't make it into the straight up default since it would have to be static (then a bunch of unclean things to make work) so I just added some overrides that call each other. Please take another look.

I also removed the default on declare for the BT navigator xml since it'll never work. Theres no reasonable absolute path to give so there's really never that's useful in.